---
title: "Honeybee KB single-cell processing"
author: "Vasilina Melnichenko & Nick Panyushev"
date: "01/06/2022"
output: html_document
---

## Environment setup

```{r setup, include=FALSE, echo=FALSE}
#BiocManager::install("DropletUtils")
library(DropletUtils)
library(Matrix)
library(tidyverse)
library(Seurat)
library(ggpointdensity)
library(scico)
library(scales)
library(data.table)
library(Matrix)
#Change the path variable to match your file hierarchy
path=("~/Paseka/Repeats_in_honeybee_genome/")
setwd(path)
source("SC_functions.R") #Contains home-made functions for scRNA analysis
```

# Creating Seurat objects

## Reading the sample sheet

```{r reading Sample sheet}
Sample_sheet <- fread("SraRunTable.txt")
```

## Load the transcript2gene table and identify mitochondrial and ribosomal genes

```{r identifying mitochondrial and ribosomal genes in tr2g and adding MT- and RB- pattern}
setwd(path)
#list of mitochondrial genes
mitoch <- fread("mitochondrial_genes", header =FALSE)
ribos <- fread("ribosomal_genes", header =FALSE)
#load the table of repeat consensi that aligned to mitochondrial genome 
mitoch_repeats <- fread("honeybee_mitochondrion_elements.csv")

mitoch$V1 <- NULL
ribos$V1 <- NULL
mitoch <- mitoch$V2
ribos <- ribos$V2
tr2g <- fread("tx2gene.tsv", sep = "\t", fill = T, header=FALSE)
colnames(tr2g) <- c("transcript", "gene_id", "gene_symbol")
tr2g[gene_id %in% mitoch, gene_id:=paste0("MT-", gene_id)]
tr2g[gene_id %in% mitoch_repeats$V1, gene_id:=paste0("MT-", gene_id)]
tr2g[gene_id %in% ribos, gene_id:=paste0("RB-", gene_id)]
tr2g[grepl("rRNA", gene_id), gene_id:=paste0("RB-", gene_id)]
tr2g$gene_symbol <- tr2g$gene_id
tr2g <- unique(tr2g[, .(gene_id, gene_symbol)])
```

## Identify mitochondrial and ribosomal genes in tx2gene

In this chunk and hereafter we will comply to standard Seurat pipeline, but optimized for keeping RAM clear from objects not operated on. That's why here are lots of lapply statements.

```{r adding "MT-" pattern to all the mitochondrial genes in cells_x_genes matrix from kb}
for (i in kb_paths){
  file_to_write <- paste("~/Paseka/", i, "/counts_unfiltered/cells_x_genes.genes.txt", sep = "")
  fwrite(mark_rb_mt_genes(file_to_write, mitoch, ribos), file = file_to_write, col.names = FALSE)
  rm(file_to_write, i)
}
```

## Load raw data from Kallisto Bustools matrices

```{r gunzip all the matrices, engine='bash', eval=FALSE}
#Unpackng all the archives with .mtx files
cd ~/Paseka/
pigz -dk -p 8 ./SRR*/counts_unfiltered/*.gz
```

```{r Load raw data, eval=TRUE}
kb_paths <- dir("../")[-1] # Do not need the first folder, as it is the git repo
                           #  itself
```

## Read KB matrices and save them as .RData objects

```{r Read_matrices, eval=FALSE}
SRRs <- sub("_out", "", kb_paths)

for (i in SRRs){
  dir <- paste("~/Paseka/", i, "_out/counts_unfiltered", sep = "")
  if(!exists(i)){ 
    assign(i, read_count_output(dir, name = "cells_x_genes"))
  }else{
        warning("Object ", i, " exists, saving on disk and removing from workspace")
  }
  save(list = i, file = paste0(i, ".RData"))
  remove(list=c(dir, i))
  gc(verbose = F, full = F)
}
rm(i)
```

## Get summary statistics of the each unfiltered dataset

```{r sum_stat_unfiltered}
counts_stat <- sapply(SRRs, function(x){
  load(paste0(x,".RData"))
  tot_counts <- Matrix::colSums(get(x)[[1]])
  return(c(dim(get(x)[[1]]), summary(tot_counts)))
})

counts_stat <- as.data.table(t(counts_stat), keep.rownames = T)
setnames(counts_stat, old = c("rn", "V1", "V2"),
         new = c("Sample", "genes_number", "cell_number"))
```

# QC

### Knee plots

```{r knee plots}
#The barcodeRanks object is around 4.8GB, be careful! 
# Executing takes ~10mins 
barcodeRanks <- sapply(SRRs, function(x){
  load(paste0(x,".RData"))
  return(barcodeRanks(get(x)[[1]], lower = 10))
})

#knee_plots <- lapply(barcodeRanks, knee_plot) - uncomment to see knee plots in here
sapply(names(knee_plots), function(x) ggsave(paste0("./Knee_plots/", x, ".png"),
  knee_plots[[x]] + ggtitle(x)))
rm(barcodeRanks)
```

### Filtering empty droplets out

```{r knee_filtering}
tmp <- lapply(SRRs, function(x){
  load(paste0(x,".RData"))
  assign(x, knee_filter(get(x)[[1]]))
  save(list = x, file = paste0("./knee_filtered/", x, ".RData"))
  })
rm(tmp)
```

### Get summary statistics after filtering

```{r sum_stat_filtered}
counts_stat_filtered <- sapply(SRRs, function(x){
  load(paste0("./knee_filtered/", x,".RData"))
  tot_counts <- Matrix::colSums(get(x))
  return(c(dim(get(x)), summary(tot_counts)))
})

counts_stat_filtered <- as.data.table(t(counts_stat_filtered), keep.rownames = T)
setnames(counts_stat_filtered, old = c("rn", "V1", "V2"),
         new = c("Sample", "genes_number", "cell_number"))
fwrite(counts_stat_filtered, "filtered_stats.tsv")
```

## Seurat object creation

```{r creating Seurat objects}
tmp <- sapply(Sample_sheet$Run,  function(x){
  load(paste0("./knee_filtered/", x,".RData"))
  assign(x, CreateSeuratObject(get(x), min.cells = 3, min.features = 200))
  save(list = x, file = paste0("./Seurats/", x, ".RData"))
  })
rm(tmp)
```

## Percentage of mitochondrial and ribosomal genes

```{r mt and rb features}
tmp <- sapply(Sample_sheet$Run,  function(x){
  ds_name <- filename <- x
  load(paste0("./Seurats/", x,".RData"))
  x <- get(x)
  x[["mito.fraction"]] <-  PercentageFeatureSet(x, pattern = "MT-")
  x[["ribo.fraction"]] <-  PercentageFeatureSet(x, pattern = "RB-")
  x[["organ"]] <- "brain"
  x[["stage"]] <- "adult"
  x[["SRA"]] <- filename
  x[["caste"]] <- unique(Sample_sheet[Run==filename, caste])
  assign(ds_name, x)
  save(list = ds_name, file = paste0("./Seurats/", filename, ".RData"))
  })
rm(tmp)
```

## Visualize QC metrics as a violin plot

```{r}
options(repr.plot.width=12, repr.plot.height=6)
mito_ribo_plots <- lapply(Sample_sheet$Run, function(x){
  sra <- x 
  load(paste0("./Seurats/tmp/", x, ".RData"))
  tmp <- VlnPlot(x, c("mito.fraction", "ribo.fraction"), ncol = 2) + 
  patchwork::plot_annotation(title = sra)
  return(tmp)     
})

#lapply(mito_ribo_plots, plot) #- uncomment to see knee plots in here

names(mito_ribo_plots) <- Sample_sheet$Run
# This lapply saves plots in MtRb_plots folder
tmp <- sapply(names(mito_ribo_plots), function(x) ggsave(paste0("./MtRb_plots/", x, ".png"),
 mito_ribo_plots[[x]]))
rm(tmp)
```

### Gather some stats about Rb and Mt percentage

```{r Mt_RB_stats}
mtrb_counts_stat <- sapply(Sample_sheet$Run, function(x){
  load(paste0("./Seurats/", x,".RData"))
  return(c(summary(get(x)@meta.data$mito.fraction),
           summary(get(x)@meta.data$ribo.fraction)))
})

mtrb_counts_stat <- as.data.table(t(mtrb_counts_stat), keep.rownames = T)

names(mtrb_counts_stat) <- c("Run", "MT-min", "MT-Q1", "MT-Median", "MT-Mean", "MT-Q3", "MT-max",
                             "RB-min", "RB-Q1", "RB-Median", "RB-Mean", "RB-Q3", "RB-max")
```

### Filter dead cells out

```{r filter cells with more than 5% mt or 30% rb}
tmp <- sapply(Sample_sheet$Run,  function(x){
  ds_name <- filename <- x
  load(paste0("./Seurats/", x,".RData"))
  x <- get(x)
  x <- subset(x, subset = mito.fraction < 5)
  x <- subset(x, subset = ribo.fraction < 30)
  assign(ds_name, x)
  save(list = ds_name, file = paste0("./Seurats/filtered/", filename, ".RData"))
  })
rm(tmp)
gc(verbose = F)
```

## Gather stats about filtered datasets

```{r}
mtrb_counts_stat_filtered <- sapply(Sample_sheet$Run, function(x){
  load(paste0("./Seurats/filtered/", x,".RData"))
  return(c(summary(get(x)@meta.data$mito.fraction),
           summary(get(x)@meta.data$ribo.fraction),
           dim(get(x))))
})

mtrb_counts_stat_filtered <- as.data.table(t(mtrb_counts_stat_filtered), keep.rownames = T)

names(mtrb_counts_stat_filtered) <- c("Run", "MT-min", "MT-Q1", "MT-Median", "MT-Mean", "MT-Q3", "MT-max",
                             "RB-min", "RB-Q1", "RB-Median", "RB-Mean", "RB-Q3", "RB-max", 
                             "n_genes", "n_cells")
fwrite(mtrb_counts_stat_filtered, file = "filtered_stats.tsv", sep = "\t")

sum(mtrb_counts_stat_filtered$n_cells) #1.2 Mln of cells
```

# Integrating all datasets

HOWTO is taken from here: <https://satijalab.org/seurat/articles/integration_large_datasets.html>

```{r normalization}
# options(future.globals.maxSize= 8912896000)
tmp <- sapply(Sample_sheet$Run,  function(x){
  ds_name <- filename <- x
  cat("Loading and normalizing ", filename, "\n")
  load(paste0("./Seurats/filtered/", x,".RData"))
  x <- get(x)
  x <- NormalizeData(x, verbose = F)
  x <- FindVariableFeatures(x, verbose = F)
  assign(ds_name, x)
  save(list = ds_name, file = paste0("./Seurats/filtered/", filename, ".RData"))
  cat(filename, " saved! \n")
  })
rm(tmp)
gc()
```

```{r getting variable features}
#read all the Seurats from disk to seurat list
all_SRRs <- lapply(Sample_sheet$Run, function(x){
  load(paste0("./Seurats/filtered/", x,".RData"))
       return(get(x))})

names(all_SRRs) <- Sample_sheet$Run

features <- SelectIntegrationFeatures(object.list = all_SRRs, nfeatures = 3000)
save(features, file = "features2integrate.RData")
rm(all_SRRs)
gc(verbose = F)
```

```{r Scaling and PCA on each ds separately}
load("features2integrate.RData")
tmp <- sapply(Sample_sheet$Run,  function(x){
  ds_name <- filename <- x
  load(paste0("./Seurats/filtered/", x,".RData"))
  x <- get(x)
  cat("Scaling ", filename, "\n")
  x <- ScaleData(x, features = features, verbose = FALSE, block.size = 500)
  cat("Running PCA on ", filename, "\n")
  x <- RunPCA(x, features = features, verbose = FALSE)
  assign(ds_name, x)
  save(list = ds_name, file = paste0("./Seurats/filtered/", filename, ".RData"))
  cat(filename, " saved! \n")
  })
rm(tmp)
gc()
```

```{r saving list of objects}
#read all the Seurats from disk to seurat list
all_SRRs <- lapply(Sample_sheet$Run, function(x){
  load(paste0("./Seurats/filtered/", x,".RData"))
       return(get(x))})

save(all_SRRs, file = "./Seurats/filtered/all_SRRs.RData")
```

```{r finding anchors for integration}
load("./Seurats/filtered/all_SRRs.RData")

anchors <- FindIntegrationAnchors(object.list = all_SRRs, reference = c(1, 2, 30, 31, 16, 17),
                                  reduction = "rpca", dims = 1:50)
gc()
save(anchors, file = "./Seurats/filtered/anchors.RData")
rm(all_SRRs)
gc()
```

All subsequent steps were performed on high-performance cluster using the separate script Integrate.R

Plots and cluster annotation will be made with Cluster_annotation.Rmd
